### ISSUE TITLE
Gameplay: Monkey Madness enhancements

### ISSUE THREAD

**Issue by Thedragon005** [Thedragon005/Archipelago-Ape-Escape#50]
### What feature would you like to see?
(IHNN edit - original title was Direct Warp to Specter 1 if in GO MODE)

Zoë in discord mentionned that it could be great to have a direct warp to Specter 1 in the minigame lobby when you have all requirements.
I don't know if it relate to the experience we want to give the players.
Some games have some steps to do to get to the boss and with Specter 1 it feels okay to clear some steps before going into the fight.

In Glitchless at most you get to do Coaster/Haunted, Circus, Go Karz, Castle Outside, Monkey Head button, Inside/Outside Climb and then Specter 1

Matter to discussion... 😀

**Comment by gamefreq0** [comment]
I think the possibility to have Natalie, the Professor, and Jake, as items and location, as mentioned in another issue here, would offer quite a bit in the way towards realizing this in a way that's opt-in; the only further thing that would help (without a direct warp), would be implementing the monkey lamps in some fashion, for entering the castle without detouring to trigger, and subsequently catching, the UFO monkeys at the front gate, and possibly pre-opening the Specter painting inside the castle. I think, then, if the first part of that is addressed, then pre-opening said painting could be its own option, to allow a short go-mode, that isn't just outright free.

**Comment by Thedragon005** [comment]
Good idea. The Go-Mode would require a bit nore items but could ne nicer to rush for if needed.

Being in go mode would mean having these items:
- MM Nathalie,Professor and Jake (Potential to only make it 1 Item as "Monkey Madness Lobby Door" but 3 checks anyway)
- "Castle Ouside Monkey Lamp"
-  Enough keys/future unlock token items
 Then it opens the painting if all is met

**Comment by jeffreybardon** [comment]
I'm all for the door to the second half of Monkey Madness becoming an item rather than using the standard method of unlocking, and making the missing bosses (Clown for Professor, and Jake for... Jake) into locations.
Natalie doesn't need to be a location as this is directly tied to catching monkeys - there are effectively already 3 locations here.
Monkey lamps being items is something I think should be its own option that's unrelated to this: the UFO section isn't particularly long, anyway, so this doesn't have much impact on go mode.
Pre-opening the painting is an interesting idea, but I think there's a better solution that we'll need for ER support anyway: for any place where the player can change the state of a different room in the level, only for that visit to the level, this changes should become persistent between visits. Off the top of my head, this would be the Crumbling Castle boss door button, Monkey Madness electric fence button, and... Monkey Madness Specter painting button.

With these few things implemented, the endgame of a seed has a lot more user flexibility in the following ways:
Final goal - tokens only, Specter 1 or Specter 2 (with or without tokens)
With level shuffle on, the climb would still have to be done at some point, but this point could be before go mode: similar to killing Agahnim 2 while checking the top of Ganon's Tower in LTTP
With level shuffle off or MM forced last, the second half door being an item still removes half of the level in a go mode situation
With ER on, the boss room could be anywhere in the level - go mode becomes scouting (and I suspect would be faster scouting, as sometimes the "climb" rooms would become "descent" rooms)

Pre-opening the painting then seems a bit overkill - I would view the MM last Specter 1 goal as one that wants a longer go mode, and it would already be much shorter than the admittedly quite long full MM experience. And if someone really wants to bypass the climb, that's what the IJ/super flyer entrance is for :P
Similarly, a direct warp also seems overkill for the same reasons.

**Comment by jeffreybardon** [comment]
Summary of the above:

- Door to second half as progression item always
- Pre-open Jake entrance?
- Clown boss and Jake boss as locations always
- Make cross-room level changes persist between visits (like opening the painting)
- Ability to block starting Specter 1 fight as in #11 (I will look into this when I have time)

**Comment by Thedragon005** [comment]
A related feature is in testing:
The 8 Monkey Lamps items (Including MM UFO Monkey Lamp) are currently in test and seem to work well.

We still need to adjust some rules being too heavily vanilla coded to go to a "Per Room" (And even per Entrance) model. Some rules are too restrictive, like Crumbling Castle.

For the MM Door all is working prefectly at blocking the door with the Monkey Lamp option.

**Comment by Thedragon005** [comment]
I can dig in bullet node 1 to 3 of IHNN's summary:
(In this order of implementation, since finding the first would probably lead to finding some usefull values for the next point)

- Clown boss and Jake boss as locations always
- Pre-open Jake entrance?
- Door to second half as progression item always

**Comment by Thedragon005** [comment]
Did a basic implementation for some of the functionalities in ``apeescape-LocalRooms`` branch:

- Door to second half as progression item always -> Item added, Client side prevention needs checking because it is tied to the Double_Door opening
- Pre-open Jake entrance? -> Needs more testing since the door have a VERY peculiar behavior
- Clown boss and Jake boss as locations always -> WORKS,locations added
- Make cross-room level changes persist between visits (like opening the painting) -> WORKS

**Comment by Thedragon005** [comment]
Current bug (Made a separate issue to track it) #55 :
Nathalie cutscenes does not play while you got the door open.

Could mean we would need to reset the state of Jake_Defeated back to 0 while in Haunted House/Nathalie Room

No impact on locations accessibility or logic, it only prevents you to exit with the Coaster Car after catching the 3 monkeys

**Comment by Thedragon005** [comment]
All is solved in the 0.7.3 release.
Only bullet point not taken into account is the ability to block S1 from being trigerred but we already have all it takes to do it. Will be done in #11 (Alternate Goal unlock condition)
